File: lab6/python/graph.py
import sys

import general

# Marker for invalid node.
INVALID_NODE = sys.maxsize

# Graph Data Structure

# def scan_node(f):

# def scan_raw_weight(f):

# def scan_weight(f):

##
 # Edge out from a node.
 #
 # Each successor is represented by a weight and a target node.
 #/
class edge_tgt_t:
    def __init__(self, w, v):
        self.w = w
        self.v = v

    def _tgt_is_invalid(self):
        return self.v == INVALID_NODE
            
class succs_t:
    def __init__(self):
        self.n = 0
        self.capacity = 0
        self.a = []
        
    def succs_add(self, tgt):
        if (self.n >= self.capacity):
            self.capacity = (self.capacity + 1)*2
            aa = [None]*self.capacity
            for i in range(0, self.n):
                aa[i] = self.a[i]
            self.a = aa
        assert(self.n < self.capacity)
        self.a[self.n] = tgt
        self.n = self.n + 1

# Class for graphs.
class graph_t:
    def __init__(self, num_nodes):
        assert(num_nodes < sys.maxsize)
        self.num_nodes = num_nodes
        if (not num_nodes == None):
            self.adjs = [None]*num_nodes
        else:
            self.adjs = None
        for i in range(0, num_nodes):
            self.adjs[i] = succs_t()
            
    def graph_check_node(self, u):
        if (not u < self.num_nodes):
            general.error("Node out of range: %d", u)

    ##
     # Get number of nodes.
     #/
    def graph_get_num_nodes(self):
        return self.num_nodes

    ##
    # Add edge to graph
    # @pre Nodes must be in range, i.e., u,v < num_nodes
    # @pre Weight must be finite
    # @pre Self-loop edges not allowed, i.e., u != v
    #/
    def graph_add_edge(self, u, w, v):
        assert (not (u == v)), "self loop edge %d -> %d" % (u, v)
        assert(w.weight_is_finite())
        
        self.graph_check_node(u)
        self.graph_check_node(v)
        
        self.adjs[u].succs_add(edge_tgt_t(w, v))

    ##
     # Get sucessors of specified node
     #
     # Each successor is represented by a weight and a target node.
     # One can query the number of successors, and array of successors
     #
     # For example, the following code snippet iterates over all successors of node u in graph g
     #
     #      for tgt in g.get_graph_succs(u):
     #          # use tgt.v and tgt.w to access target node and edge weight 
     #
     #/
    def graph_num_succs(self, u):
        return self.adjs[u].n
        
    def get_graph_succs(self, i):
        return self.adjs[i].a[0:self.adjs[i].n]
        
    # Create graph from file. The file is a text file containing the positive integer numbers:
    #
    # num_nodes
    # num_edges
    # u_1 v_1 w_1
    # ...
    #
    # i.e., the number of nodes, then the number of edges, and then, for each edge, the source node, target node, and weight.
    #
    # Everything content after the last edge will be ignored.
    #
    # The function outputs an error message for files not adhering to this format!
    #/
    # graph_read - not implemented since not needed for practcal

    ##
     # Write graph to file. Uses format described in @ref graph_read.
     #/
    def graph_write(self, f):
        N = self.graph_get_num_nodes()
        num_edges = 0
        
        for u in range(0, N):
            num_edges += self.graph_num_succs(u)
            
        f.write("%d\n%d\n" % (N, num_edges))
        
        for u in range(0, N):
            for tgt in self.get_graph_succs(u):
                if (not tgt == None):
                    f.write("%d %d %d\n" % (u, tgt.v, tgt.w.weight_to_int()))
                else:
                    f.write("NONE\n")
#/// Paths
##
 # A path is a sequence of nodes with a specified length.
 #/
class path_t:
    def set_up(self, len):
        self.len = len
        self.nodes = [None]*len
        for i in range(0, len):
            self.nodes[i] = INVALID_NODE

    ##
     # Extract path from predecessor array. Convention: Start node points to itself.
     #/
    def __init__(self, pred, v):
        # Determine length
        len = 0
        u = v
        
        while (True):
            len = len + 1
            if (u == pred[u]):
                break
            u = pred[u]
            
        self.set_up(len)
        u = v
        
        while (True):
            len = len - 1
            self.path_set(len, u)
            if (u == pred[u]):
                break
            u = pred[u]

    def path_set(self, i, u):
        assert (i < self.len)
        self.nodes[i] = u
    
    def path_get(self, i):
        assert(i < self.len)
        return self.nodes[i]
        
    def path_len(self):
        return self.len
        
        
        

File: lab6/c/graph.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdarg.h>
#include <stdbool.h>
#include <assert.h>
#include <limits.h>
#include <math.h>

#include "general.h"
#include "graph.h"



//////////////////////// Graph Data Structure

node_t scan_node(FILE *file) {
    node_t res;
    if (fscanf(file, "%zu", &res) == 1) return res; else error("Error reading input");
}

raw_weight_t scan_raw_weight(FILE *file) {
    raw_weight_t res;
    if (fscanf(file,"%ld", &res) == 1) return res; else error("Error reading input");
}

weight_t scan_weight(FILE *file) {
    signed long res;
    if (fscanf(file,"%ld", &res) == 1) return weight_of_int(res); else error("Error reading input");
}

//const node_t INVALID_NODE = SIZE_MAX;

// const edge_tgt_t INVALID_TGT = {weight_inf(),INVALID_NODE};

bool _tgt_is_invalid(edge_tgt_t const *tgt) {
    return tgt->v == INVALID_NODE;
}


typedef struct _succs_t {
    size_t n;
    size_t capacity;   // Capacity
    edge_tgt_t *a;
} succs_t;


void succs_init(succs_t *s) {
    s->n=0;
    s->capacity=0;
    s->a = NULL;
}

void succs_uninit(succs_t *s) {
    if (s->a) free(s->a);
}

void succs_add(succs_t *s, edge_tgt_t tgt) {
    if (s->n >= s->capacity) {
        s->capacity = (s->capacity+1)*2;
        edge_tgt_t *aa = CALLOC(edge_tgt_t,s->capacity);
        if (s->a) {
            memcpy(aa,s->a,sizeof(edge_tgt_t)*s->n);
            free(s->a);
        }
        s->a = aa;
    }
    assert(s->n < s->capacity);
    s->a[s->n] = tgt;
    ++s->n;
}


struct _graph_t {
    size_t num_nodes;
    succs_t *adjs;
};

// Graph
void graph_init(graph_t *g, size_t num_nodes) {
    assert(num_nodes < SIZE_MAX);
    g->num_nodes=num_nodes;
    if (num_nodes) g->adjs = CALLOC(succs_t,num_nodes); else g->adjs=NULL;
    for (size_t i=0; i<num_nodes;++i) succs_init(g->adjs + i);
}

void graph_uninit(graph_t *g) {
    for (size_t i=0; i<g->num_nodes;++i) succs_uninit(g->adjs + i);
    if (g->adjs) free(g->adjs);
}

graph_t *graph_new(size_t num_nodes) {
    graph_t *res = CALLOC(graph_t,1);
    graph_init(res,num_nodes);
    return res;
}

void graph_delete(graph_t *g) {
    graph_uninit(g);
    free(g);
}

void graph_check_node(graph_t const *g, node_t u) {
    if (!(u < g->num_nodes)) {error("Node out of range: %zu",u);}
}

size_t graph_get_num_nodes(graph_t const *g) {return g->num_nodes;}


void graph_add_edge(graph_t *g, node_t u, weight_t w, node_t v) {
//     printf("Add edge %lu %lu %ld\n",u,v,w);
    assertmsg(u!=v,"self loop edge %zu -> %zu", u, v);
    assert(weight_is_finite(w));

    graph_check_node(g,u); graph_check_node(g,v);
    succs_add(g->adjs + u, (edge_tgt_t){w,v});
}

size_t graph_num_succs(graph_t const *g, node_t u) {
    return g->adjs[u].n;
}

edge_tgt_t const *graph_succs_begin(graph_t const *g, node_t u) {
    return g->adjs[u].a;
}

edge_tgt_t const *graph_succs_end(graph_t const *g, node_t u) {
    return g->adjs[u].a + g->adjs[u].n;
}

graph_t *graph_read(FILE *f) {
    size_t num_nodes = scan_size(f);
    size_t num_edges = scan_size(f);
    graph_t *g = graph_new(num_nodes);

//     printf("Reading %zu nodes and %zu edges\n",num_nodes,num_edges);

    for (size_t i=0; i<num_edges; ++i) {
        node_t u = scan_node(f);
        node_t v = scan_node(f);
        weight_t w = scan_weight(f);
        graph_add_edge(g,u,w,v);
    }

    return g;
}

void graph_write(FILE *f, graph_t const *g) {
  size_t N = graph_get_num_nodes(g);
  size_t num_edges = 0;

  for (node_t u=0; u<N; ++u) {
    num_edges+=graph_num_succs(g,u);
  }

  fprintf(f, "%zu\n%zu\n",N,num_edges);

  for (node_t u=0; u<N; ++u) {
    for (edge_tgt_t const *tgt = graph_succs_begin(g,u); tgt!=graph_succs_end(g,u);++tgt) {
      fprintf(f, "%zu %zu %ld\n",u,tgt->v,weight_to_int(tgt->w));
    }
  }

}



struct _path_t {
  size_t len;
  node_t nodes[];
};

path_t *path_new(size_t len) {
  path_t *p = malloc(sizeof(path_t) + len * sizeof(node_t));
  if (!p) error("Out of memory");
  p->len=len;
  for (size_t i=0;i<len;++i) p->nodes[i]=INVALID_NODE;
  return p;
}

void path_delete(path_t *p) {
  free(p);
}

void path_set(path_t *p, size_t i, node_t u) {
  assert(i<p->len);
  p->nodes[i]=u;
}
node_t path_get(path_t const *p, size_t i) {
  assert(i<p->len);
  return p->nodes[i];
}

size_t path_len(path_t const *p) {return p->len;}

path_t *path_from_pred(node_t const *pred, node_t v) {
  // Determine length
  size_t len=0;

  node_t u=v;

  while (true) {
    ++len;
    if (u==pred[u]) break;
    u=pred[u];
  }
  path_t *p = path_new(len);

  u=v;
  while (true) {
    --len;
    path_set(p,len,u);
    if (u==pred[u]) break;
    u=pred[u];
  }

  return p;
}







